### DISCUSSION TITLE
Using localeCompare to sort posts in foreign language

### DISCUSSION THREAD

**Discussion by kaviars** [11ty/eleventy#1732] [Q&A]
Dear community,

Somehow sorting posts in Latvian results in wrong order. Please help to get the posts sorted by alphabet.
Alphabet (in total Latvian alphabet consits of 33 letters) first 8 letters:
A a | Ā ā | B b | C c | Č č | D d | E e | Ē ē ...

To sort posts according to Latvian alphabet I use localeCompare as follows:
```
  eleventyConfig.addCollection("postsDescending", function(collection) {
    return collection.getFilteredByGlob("src/posts/*.md").sort(function(a, b) {
      return a.data.title.localeCompare(b.data.title, "lv-LV", {
        sensitivity: "variant",
        ignorePunctuation: true,
        usage: "sort"
      });
    });
  });
```
In result posts starting with long variation letter are displayed before posts starting with short variation letter, but should be the opposite. 

Result:
1. Ērenpress ir velosipēdu ražotājs
2. Egles zarā sēž kaķis

Should be:
1. Egles zarā sēž kaķis
2. Ērenpress ir velosipēdu ražotājs

Just to be sure if localeCompare actually works, tried with other languages and sorting is fine, but not for my native Latvian.
Any ideas how to force sorting posts according to Latvian alphabet?

Tested localeCompare without Eleventy and it does work. #confused
```
"Ā".localeCompare("a", "lv-LV", {
  caseFirst: "upper",
  sensitivity: "variant",
  usage: "sort"
}) // 1
```

Thanks in advance,

**Comment by kaviars** [comment]
Alarm! Beginner in modern web dev!
Perhaps [localeCompare polyfill](https://github.com/jeppeburchardt/locale-compare-polyfill) is necessary to get it working properly? Not sure how to set it up though.. in `.eleventy.js` I've added `const localeComparePolyfill = require("locale-compare-polyfill");`, but how to connect to my "postsDescending" collection? Swapping `localeCompare` (from       `return a.data.title.localeCompare(b.data.title, "lv", {`
) with `localeComparePolyfill` results in error.

**Reply by pdehaan** [reply]
I don't think you should need a polyfill if you're running this in Node and not directly in an old browser.

I tried making a simple test case (without 11ty) and I think I'm seeing your expected results on macOS 10.15.7 w/ Node 14.16.0:

```js
const titles = [
  "Ērenpress ir velosipēdu ražotājs",
  "Egles zarā sēž kaķis"
];

console.log(titles.sort((a, b) => {
  return a.localeCompare(b, "lv-LV", {
    caseFirst: false,
    sensitivity: "variant",
    ignorePunctuation: true,
    usage: "sort"
  });
}).join("\n"));

/*
Egles zarā sēž kaķis
Ērenpress ir velosipēdu ražotājs
*/
```

Even if I add a [basic 11ty blog](https://github.com/pdehaan/locale-compare-latvia) w/ 4 posts and an index page, it seems to render the collection in the order you were expecting:

```html
<h1>HOMEPAGE</h1>
<h2>Posts (sorted)</h2>
<ul>
  <li>aaa</li>
  <li>Egles zarā sēž kaķis</li>
  <li>Ērenpress ir velosipēdu ražotājs</li>
  <li>ZZZ</li>
</ul>
```

**Reply by kaviars** [reply]
Thanks for your input! Much appreciated! 
Downloaded the basic 11ty blog to add few more posts. With that done, realized localeCompare takes next letter in word, despite words first letters are ordered wrong.

Below is a list with more dummy post entries:

Egles zarā sēž kaķis
Ēna
Epizode
Ērenpress ir velosipēdu ražotājs
Ēvele

 I expect "Ēna" to come after "Epizode", but seems the second letter ("n" in "Ēna")plays role in displaying word before "Epizode", because in alphabet letter "n" comes before "p"(second letter in word "Epizode"). Another same example is with word "Ēvele". If that is changed to "Evele", order does not change, because its second letter ("v") comes after "r" (in "Ērenpress").
Played around with [localeCompare options](https://www.techonthenet.com/js/string_localecompare.php), but still cant get the order right. Handy [example for sensitivity option](https://javascriptqa.com/images/1/tOaMmJTWta.png), although result does not change.. Brr.

**Reply by pdehaan** [reply]
Does this help at all? https://stackoverflow.com/questions/62142376/sort-titles-with-localcompare

**Reply by kaviars** [reply]
I am the author of that post 😬

**Reply by kaviars** [reply]
Fellow countryman helped out explaining each line to get sorting done without LocaleCompare. Though in Eleventy the code does not work, still sorting wrong :/
Working sample without Eleventy
https://jsfiddle.net/virsis12/L7vkxjun/
```javascript
  eleventyConfig.addCollection("postsDescending", function (collection) {
    const alphabet = ['A','a','Ā','ā','B','b','C','c','Č','č','D','d','E','e','Ē','ē','F','f','G','g','Ģ','ģ','H','h','I','i','Ī','ī','J','j','K','k','Ķ','ķ','L','l','Ļ','ļ','M','m','N','n','Ņ','ņ','O','o','P','p','R','r','S','s','Š','š','T','t','U','u','Ū','ū','V','v','Z','z','Ž','ž'];
    return collection.getFilteredByGlob("src/posts/*.md")
    .sort(function (a, b) {
          const byLetters = (index) => {
            const titleA = a[index], titleB = b[index];
            if (!titleA || !titleB) {
              return titleA ? 1 : -1;
            }
            if (titleA === titleB) {
              return byLetters(index+1);
            }
            const indexA = alphabet.findIndex(c => c === titleA);
            const indexB = alphabet.findIndex(c => c === titleB);
            if (indexA < indexB) {
                return -1;
            }
            if (indexA > indexB) {
                return 1;
            }
          }
          return byLetters(0);      
        });
    });
```

**Reply by pdehaan** [reply]
I _think_ your problem is around here:

```js
const titleA = a[index], titleB = b[index];
```

... since `a` and `b` are page objects here and not simple strings (like the vanilla jsfiddle code). You probably want something like this:

```js
    const titleA = a.data.title[index];
    const titleB = b.data.title[index];
```

It seems to work locally for me, but I should add more pages like your sample Latvian strings in 
```
const titles = [
  "Ēnas",
  "Ēna",
  "Epizode",
  "Ēvele",
  "Egles zarā sēž kaķis",
  "Ērenpress ir velosipēdu ražotājs",
];
```

**Reply by kaviars** [reply]
🤦 How did I miss that ... Thanks for staying till end! 🙏🙏

**Reply by pdehaan** [reply]
And here's what my archive page looks like for me in 11ty w/ more pages:

```html
<h2>Posts (sorted)</h2>
<ul>
  <li>aaa</li>
  <li>Egles zarā sēž kaķis</li>
  <li>Egles zarā sēž kaķis</li>
  <li>Egles zarā sēž kaķis</li>
  <li>Epizode</li>
  <li>Epizode</li>
  <li>Ēna</li>
  <li>Ēnas</li>
  <li>Ērenpress ir velosipēdu ražotājs</li>
  <li>Ērenpress ir velosipēdu ražotājs</li>
  <li>Ēvele</li>
  <li>ZZZ</li>
</ul>
```
